feat: Add XMP metadata buffer support to withMetadata()

tpatel opened the pull request:
- Add xmp parameter to withMetadata() function to accept Buffer containing XMP metadata
- Implement native C++ handling for XMP buffer processing using VIPS_META_XMP_NAME
- Add test suite with 13 test cases covering XMP functionality
- Support XMP metadata for PNG, JPEG, WebP, and TIFF formats
- Include proper Buffer validation and memory management
- Update TypeScript definitions to include xmp parameter

The implementation allows users to inject custom XMP metadata into images:
sharp(input).withMetadata({ xmp: xmpBuffer }).png().toBuffer()

mbklein committed:
Expose JPEG 2000 oneshot decoder option #4262

Requires libvips compiled with support for JP2 images

Co-authored-by: Kleis Auke Wolthuizen <github@kleisauke.nl>

tpatel committed:
feat: Add withXmp() method for XMP metadata injection

partially resolves #4143

- Add withXmp(xmpBuffer) method to embed custom XMP metadata in output images
- Include parameter validation for non-Buffer inputs
- Add tests
- Update API documentation

tpatel commented:
this is a draft pull request for internal feedback, before contributing upstream

kleisauke committed:
Support composite op with non-sRGB pipeline colourspace (#4412)

lovell committed:
Upgrade to libvips v8.17.0

CI: Use more recent, non-deprecated Windows runners

Bump devDeps

thomastoye commented on lib/index.d.ts:
```diff
@@ -1140,6 +1140,8 @@ declare namespace sharp {
          */
         exif?: Exif | undefined;
+        /** Buffer containing XMP metadata to be embedded in the output image. */
+        xmp?: Buffer | undefined;
```
The `exif` field above this has `@deprecated Use `withExif()` or `withExifMerge()` instead.`. Perhaps the `xmp` field should be private?

tpatel replied:
good point, I've started refactoring away from this method to a `withXmp` method instead 👍

tpatel replied:
addressed in https://github.com/tpatel/sharp/pull/1/commits/de3a1344d956071f7135464ec593037f7acfaef1 ✅

lovell committed:
Upgrade to sharp-libvips v1.2.0-rc.2

kleisauke committed:
Docs: update link to concurrency API (#4414)

lovell committed:
Remove prebuild dependency

Every CI build task tagged with 'package' now populates and
publishes the relevant npm/platform directory as an artefact.
These are aggregated by a fan-in release task at the end to
create a complete npm workspace zipfile. If the commit is
tagged, a release is created and the npm workspace attached.

lovell committed:
CI: Switch ARM64 from CircleCI to GitHub Actions

lovell committed:
Prerelease v0.34.3-rc.0

lovell committed:
Add Magic Kernel Sharp (no relation) to resizing kernels

lovell committed:
Expose keepDuplicateFrames GIF output parameter

lovell committed:
Nest format-specific constructor params (deprecate at top-level)

- `subifd` -> `tiff.subifd`
- `level` -> `openSlide.level`
- `pdfBackground` -> `pdf.background`

DeltaEvo commented on src/pipeline.cc:
```diff
@@ -892,6 +892,14 @@ class PipelineWorker : public Napi::AsyncWorker {
+      // XMP buffer
+      if (baton->withXmpBuffer != nullptr && baton->withXmpBufferLength > 0) {
+        image = image.copy();
+        image.set(VIPS_META_XMP_NAME, reinterpret_cast<VipsCallbackFn>(vips_area_free_cb),
```
nit: Reusing `vips_area_free_cb` doesn't feel right here, the current implementation do the thing that we want https://github.com/libvips/libvips/blob/594187812907bca9499205b8d2e0e35d724f5e3f/libvips/iofuncs/type.c#L137

`g_free` of `mem` without doing anything with the `area`, but nothing guarantees that it will do it in the future. With the size of the implementation I would provide our own free callback (with the added benefit that you will be able to remove the `reinterpret_cast`)

tpatel replied:
the goal is to contribute upstream, so I would assume it's ok to reuse something similar to https://github.com/tpatel/sharp/blob/90bc17db1ab618cdd61e9f028ee99c19710d129f/src/common.cc#L593

DeltaEvo commented on src/pipeline.h:
```diff
@@ -201,6 +201,8 @@ struct PipelineBaton {
   std::string withIccProfile;
   std::unordered_map<std::string, std::string> withExif;
   bool withExifMerge;
+  void *withXmpBuffer;
```
nit: You can use a `unique_ptr<uint8_t[]>` here, this would allow to let c++ handle the deletion on "error" cases (remove the handling in OnOK) and you have a [release](https://en.cppreference.com/w/cpp/memory/unique_ptr/release.html) function on it to pass the memory to "legacy" apis like VipsImage#set

DeltaEvo commented on src/pipeline.cc:
```diff
@@ -892,6 +892,14 @@ class PipelineWorker : public Napi::AsyncWorker {
         }
       }
+      // XMP buffer
+      if (baton->withXmpBuffer != nullptr && baton->withXmpBufferLength > 0) {
```
nit: Should it be wrapped in a `baton->keepMetadata & VIPS_FOREIGN_KEEP_XMP`

tpatel replied:
fixed in https://github.com/tpatel/sharp/commit/90bc17db1ab618cdd61e9f028ee99c19710d129f

DeltaEvo commented on src/pipeline.cc:
```diff
@@ -892,6 +892,14 @@ class PipelineWorker : public Napi::AsyncWorker {
       }
+      // XMP buffer
+      if (baton->withXmpBuffer != nullptr && baton->withXmpBufferLength > 0) {
+        image = image.copy();
```
very very small nit here since there is the same for exif, but I don't understand why we copy the image here 🤔

DeltaEvo commented on lib/output.js:
```diff
@@ -350,9 +350,17 @@ function keepMetadata () {
  * @param {Object} [options]
  * @param {number} [options.orientation] Used to update the EXIF `Orientation` tag, integer between 1 and 8.
  * @param {number} [options.density] Number of pixels per inch (DPI).
+ * @param {Buffer} [options.xmp] Buffer containing XMP metadata to be embedded in the output image.
```
nit: Allow it to be a Uint8Array ? (This is a Buffer subclass)

lovell committed:
Expose stylesheet and highBitdepth SVG input params

lovell committed:
Docs: expand info about parallelism and concurrency control #4411

kleisauke committed:
Ensure SVG scale-on-load optimisation uses newly exposed params (#4415)

Follow-up to commit c4b1d80.
